### PR TITLE
Refactor Address and HostPort from network to core/network

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Refactor Address, HostPort and related code from network to core/network.
Part of network improvements work.  

Remove unused functions: SelectPublicHostPort, SelectInternalHostPort, 
LoopbackIPv6CIDR, LoopbackIPv4CIDR, ResolvableHostnames, 
ResolveOrDropHostnames,

## QA steps

Unit tests.  There should be no impact to current juju functionality.

### CODE REVIEW

File: core/network/hostport.go [hunk 1]
```diff
@@ -131,49 +131,6 @@ func SortHostPorts(hps []HostPort) {
...
 
-var netLookupIP = net.LookupIP
-
-// ResolveOrDropHostnames tries to resolve each address of type
-// HostName (except for "localhost" - it's kept unchanged) using the
-// local resolver. If successful, each IP address corresponding to the
-// hostname is inserted in the same order. If not successful, a debug
-// log is added and the hostname is removed from the list. Duplicated
-// addresses after the resolving is done are removed.
-func ResolveOrDropHostnames(hps []HostPort) []HostPort {
```
Comment (hmlanigan) [inline core/network/hostport.go]: drive by: Unused outside of being tested.

**Comment by hmlanigan** [comment]
!!build!!

File: core/network/address.go [hunk 2]
```diff
@@ -34,3 +32,624 @@ var GenerateVirtualMACAddress = func() string {
...
+
+	// Retrieve the indexes of the addresses with the best scope and type match.
+	allowedMatchTypes := []scopeMatch{exactScopeIPv4, exactScope, firstFallbackScopeIPv4, firstFallbackScope, secondFallbackScopeIPv4, secondFallbackScope}
+	for _, matchType := range allowedMatchTypes {
+		indexes, ok := matches[matchType]
+		if ok && len(indexes) > 0 {
+			return indexes
+		}
+	}
+	return []int{}
```
Comment (SimonRichardson) [inline core/network/address.go]: Tiny nit:

```suggestion
	return make([]int, 0)
```

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**